File: p647m/palindromic_substring.js
/**
 * @param {string} s
 * @return {number}
 */
var countSubstrings = function (s) {
  let [ans, slen] = [0, s.length];
  const checkPalindrome = (l, r) => {
    while (l >= 0 && r < slen && s[l] === s[r]) {
      [l, r, ans] = [l - 1, r + 1, ans + 1];
    }
  };
  for (let i = 0; i < slen; ++i) {
    checkPalindrome(i, i); // check odd len palindrome
    checkPalindrome(i, i + 1); // check odd len palindrome
  }
  return ans;
};
// TEST
const f = (s, expected) => [s, expected];
[
  f('aa', 3),
  f('ab', 2),
  f('aab', 4),
  f('aba', 4),
  f('aaa', 6),
  f('aaaa', 10),
].forEach(([s, expected]) => {
  const actual = countSubstrings(s);
  console.log('# of palindromic substrings in', s, '->', actual);
  console.assert(actual === expected);
});

File: p647m/palindromic_substring.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        ans = 0

        def count(l: int, r: int) -> None:
            nonlocal s, ans
            while l >= 0 and r < len(s) and s[l] == s[r]:
                l, r, ans = l - 1, r + 1, ans + 1

        for i, _ in enumerate(s):
            count(i, i)
            count(i, i + 1)
        return ans


# TESTS
for s, expected in [
    ("aa", 3),
    ("ab", 2),
    ("aab", 4),
    ("aba", 4),
    ("aaa", 6),
    ("aaaa", 10),
]:
    sol = Solution()
    actual = sol.countSubstrings(s)
    print("# of palindromic substrings in", s, "->", actual)
    assert actual == expected
